### PR TITLE
Enable mixer on ingress and egress

### PR DESCRIPTION
**Pull request by kyessenov**

### CODE REVIEW

**Comment by rshriram** [comment]
Question: if we can derive the source and destination attributes dynamically , can we invoke the mixer on outbound path for calls to external services from sidecar ?

File: proxy/context.go [hunk 1]
```diff
@@ -44,84 +44,68 @@ type Environment struct {
...
 	return out, nil
 }
 
 const (
-	// EgressNode is the service node for egress proxies
-	EgressNode = "egress"
-
-	// IngressNode is the service node for ingress proxies
-	IngressNode = "ingress"
+	serviceNodeSeparator = "~"
```
Comment (rshriram) [inline proxy/context.go]: Does this work with Envoy ? And will Prometheus or other systems throw a tantrum if they see a ~ ?, any reason for omitting colon?

File: proxy/envoy/config.go [hunk 2]
```diff
@@ -136,32 +137,59 @@ func buildConfig(listeners Listeners, clusters Clusters, lds bool, mesh *proxyco
...
+	case proxy.Egress:
+		return buildEgressListeners(env.Mesh, role)
+	}
+	return nil
+}
+
+func buildClusters(env proxy.Environment, role proxy.Role) (clusters Clusters) {
+	switch role.Type {
+	case proxy.Sidecar:
+		_, clusters = buildSidecar(env, role)
```
Comment (rshriram) [inline proxy/envoy/config.go]: Same function names?

File: proxy/envoy/config.go [hunk 3]
```diff
@@ -136,32 +137,59 @@ func buildConfig(listeners Listeners, clusters Clusters, lds bool, mesh *proxyco
 	return out
 }
 
-// buildListeners produces a list of listeners and referenced clusters.
-func buildListeners(env proxy.Environment, sidecar proxy.Sidecar) (Listeners, Clusters) {
+// buildListeners produces a list of listeners and referenced clusters for all proxies
+func buildListeners(env proxy.Environment, role proxy.Role) Listeners {
+	switch role.Type {
+	case proxy.Sidecar:
+		listeners, _ := buildSidecar(env, role)
```
Comment (rshriram) [inline proxy/envoy/config.go]: For clarity, maybe be use  buildSidecarListeners ?

**Comment by kyessenov** [comment]
We can, but mixer would panic since it expects "target.service". For outbound calls to external services, there is no cluster target service.

File: proxy/envoy/config.go (hunk 2 above)
Comment (rshriram) [inline proxy/envoy/config.go]: Okay I see the difference. But this is bit unclean. Not to mention computing listener and cluster twice.

File: proxy/envoy/config.go [hunk 4]
```diff
@@ -136,32 +137,59 @@ func buildConfig(listeners Listeners, clusters Clusters, lds bool, mesh *proxyco
...
+		_, clusters = buildSidecar(env, role)
+	case proxy.Ingress:
+		httpRouteConfigs, _ := buildIngressRoutes(env.Mesh, env.ServiceDiscovery, env.IstioConfigStore)
+		clusters = httpRouteConfigs.clusters().normalize()
+	case proxy.Egress:
+		httpRouteConfigs := buildEgressRoutes(env.Mesh, env.ServiceDiscovery)
+		clusters = httpRouteConfigs.clusters().normalize()
+	}
+
+	// apply custom policies for outbound clusters
```
Comment (rshriram) [inline proxy/envoy/config.go]: You mean destination policy ?

File: proxy/envoy/config.go [hunk 5]
```diff
@@ -136,32 +137,59 @@ func buildConfig(listeners Listeners, clusters Clusters, lds bool, mesh *proxyco
...
+		clusters = append(clusters, buildMixerCluster(env.Mesh))
+	}
+
+	return clusters
+}
+
+// buildSidecar produces a list of listeners and referenced clusters for sidecar proxies
+// TODO: this implementation is inefficient as it is recomputing all the routes for all proxies
+// There is a lot of potential to cache and reuse cluster definitions across proxies and also
+// skip computing the actual HTTP routes
```
Comment (rshriram) [inline proxy/envoy/config.go]: Thanks. As long as its todo, we can tackle this.

File: proxy/envoy/config.go [hunk 6]
```diff
@@ -256,13 +315,14 @@ func buildTCPListener(tcpConfig *TCPRouteConfig, ip string, port int) *Listener
 }
 
 // buildOutboundListeners combines HTTP routes and TCP listeners
-func buildOutboundListeners(instances []*model.ServiceInstance, services []*model.Service,
-	config model.IstioConfigStore, mesh *proxyconfig.ProxyMeshConfig) (Listeners, Clusters) {
+func buildOutboundListeners(mesh *proxyconfig.ProxyMeshConfig, sidecar proxy.Role, instances []*model.ServiceInstance,
```
Comment (rshriram) [inline proxy/envoy/config.go]: s/sidecar/role/ ?

File: proxy/context.go [hunk 7]
```diff
@@ -44,84 +44,68 @@ type Environment struct {
 	Mesh *proxyconfig.ProxyMeshConfig
 }
 
-// Role declares the proxy node role in the mesh
-type Role interface {
-	// nolint: megacheck
-	isProxyRole()
+// Role defines the proxy role
+type Role struct {
```
Comment (rshriram) [inline proxy/context.go]: This is overloading the term with more info. Suggest using a different noun like router or ProxyAttributes or something

File: proxy/envoy/config.go (hunk 6 above)
Comment (rshriram) [inline proxy/envoy/config.go]: Ignore. I commented above about renaming proxy.role to attributes or something

File: proxy/envoy/egress.go [hunk 8]
```diff
@@ -42,20 +50,26 @@ func buildEgressRoutes(services model.ServiceDiscovery, mesh *proxyconfig.ProxyM
...
+			}
 
-			// overwrite cluster hosts and types
-			cluster.Type = ClusterTypeStrictDNS
-			cluster.Hosts = []Host{{
-				URL: fmt.Sprintf("tcp://%s:%d", svc.ExternalName, servicePort.Port),
-			}}
+			if servicePort.Protocol == model.ProtocolGRPC || servicePort.Protocol == model.ProtocolHTTP2 {
+				cluster.Features = ClusterFeatureHTTP2
+			}
```
Comment (rshriram) [inline proxy/envoy/egress.go]: If auth is enabled, I believe we need to set alpn_protocol or something to http2. Not sure if that is handled elsewhere

File: proxy/context.go (hunk 7 above)
Comment (kyessenov) [inline proxy/context.go]: is `ProxyNode` better?

File: proxy/context.go (hunk 1 above)
Comment (kyessenov) [inline proxy/context.go]: only ":" messes with statsd in envoy, other characters are fine. this also has to be a legit URI character.

File: proxy/envoy/config.go (hunk 2 above)
Comment (kyessenov) [inline proxy/envoy/config.go]: There's a big TODO about separating the two, but that's a lot of work.

File: proxy/envoy/egress.go (hunk 8 above)
Comment (kyessenov) [inline proxy/envoy/egress.go]: nope, let's get this alpn fixed soon, but in another PR for better review.

File: cmd/pilot-agent/main.go [hunk 9]
```diff
@@ -31,7 +31,7 @@ import (
 var (
 	configpath string
 	meshconfig string
-	role       proxy.Role
+	role       proxy.Node
```
Comment (rshriram) [inline cmd/pilot-agent/main.go]: nit: role->node, but its upto you.

**Comment by sebastienvas** [comment]
@istio-testing smoke test this

**Comment by kyessenov** [comment]
@istio-testing smoke test this

**Comment by kyessenov** [comment]
@istio-testing smoke test this

**Comment by kyessenov** [comment]
Test passed locally. Force merging with @douglas-reid 's blessing.

**Pull request merged**